sklearn/inspect/partial_dependence.py

```diff
@@ -0,0 +1,589 @@
+        array ``values[j]`` is either ``grid_resolution``, or the number of
+        unique values in ``X[:, j]``, whichever is smaller.
+    """
+    try:
+        assert len(percentiles) == 2
+    except (AssertionError, TypeError):
+        raise ValueError('percentiles must be a sequence of 2 elements.')
+    if not all(0. <= x <= 1. for x in percentiles):
+        raise ValueError('percentiles values must be in [0, 1].')
+    if percentiles[0] >= percentiles[1]:
+        raise ValueError('percentiles[0] must be strictly less '
+                         'than percentiles[1].')
+
+    if grid_resolution <= 1:
+        raise ValueError('grid_resolution must be strictly greater than 1.')
+
+    values = []
+    for feature in range(X.shape[1]):
+        uniques = np.unique(X[:, feature])
+        if uniques.shape[0] < grid_resolution:
+            # feature has low resolution use unique vals
+            axis = uniques
+        else:
+            # create axis based on percentiles and grid resolution
+            emp_percentiles = mquantiles(X, prob=percentiles, axis=0)
+            if np.allclose(emp_percentiles[0, feature],
+                           emp_percentiles[1, feature]):
+                raise ValueError('percentiles are too close to each other, '
+                                 'unable to build the grid.')
+            axis = np.linspace(emp_percentiles[0, feature],
+                               emp_percentiles[1, feature],
+                               num=grid_resolution, endpoint=True)
+        values.append(axis)
+
+    return cartesian(values), values
+
+
+def _partial_dependence_recursion(est, grid, features):
+    if est.init is not None:
+        warnings.warn(
+            'Using recursion method with a non-constant init predictor will '
+            'lead to incorrect partial dependence values.',
+            UserWarning
+        )
+
+    # grid needs to be DTYPE
+    grid = np.asarray(grid, dtype=DTYPE, order='C')
+
+    n_trees_per_stage = est.estimators_.shape[1]
+    n_estimators = est.estimators_.shape[0]
+    learning_rate = est.learning_rate
+    averaged_predictions = np.zeros((n_trees_per_stage, grid.shape[0]),
+                                    dtype=np.float64, order='C')
+    for stage in range(n_estimators):
+        for k in range(n_trees_per_stage):
+            tree = est.estimators_[stage, k].tree_
+            _partial_dependence_tree(tree, grid, features,
+                                     learning_rate, averaged_predictions[k])
+
+    return averaged_predictions
+
+
+def _partial_dependence_brute(est, grid, features, X):
+    averaged_predictions = []
+    for new_values in grid:
+        X_eval = X.copy()
+        for i, variable in enumerate(features):
+            X_eval[:, variable] = new_values[i]
+
+        try:
+            predictions = (est.predict(X_eval) if is_regressor(est)
+                           else est.predict_proba(X_eval))
```

Comment:
What about decision_function?

Reply:
Not every classifier has a decision function right?

Do you think this should be a parameter: plotting the result of decision_function rather than predict_proba?

Reply:
The usual pattern is to try if either exists. Not all models have predict_proba either. I'm fine with trying predict_proba first but if it doesn't exist fall back to decision_function.

Reply:
Might be best to make this explicit. Provide a parameter to choose the prediction methods, with a default Auto behaviour